process-engine/process_engine_core
# 🐛 Fix BoundaryEvent state transitions

## Changes

1. Detach BoundaryEvents prior to running `afterExecute` cleanup on decorated Activities
2. Implement resumption for BoundaryEventHandlers
3. When resuming a decorated activity, all attached BoundaryEvents will now be resumed, instead of executed
    - This fixes the issue with multiple `onEnter` tokens for a single BoundaryEvent
4. Fix some reference issues
5. Unsubscribe from `ExternalTaskFinished` event, when an `ExternalServiceTaskHandler` gets interruped
6. Fix finishing of BoundaryEvents
    - Each BoundaryEvent should now always produce an `onExit` token, either when trigger, or when the activity finishes and removes its BoundaryEvents
7. Refine resumption of ProcessInstances. A ProcessInstance is now considered to be orphaned, if it:
    - is in a running state
    - has no active FlowNodeInstances
    - has reached at least one EndEvent
8. ProcessInstances that were in a running state, have no active FlowNodeInstances, but have not yet reached an EndEvent are regarded as interrupted and will be resumed regularly.

## Issues

Part of https://github.com/process-engine/process_engine_runtime/issues/510

PR: #318

## How to test the changes

To check the BoundaryEvents:

- Run a Process that uses decorated activities
    - easiest would be a UserTask or an external ServiceTask because they do not automatically resume
- Decorate these activities with either:
    - `TimerBoundaryEvent`
    - `MessageBoundaryEvent`
    - `SignalBoundaryEvent`
- Trigger the BoundaryEvent 
    - Messages and Signals can be easily triggered through a separate process that uses IntermediateEvents or dedicated EndEvents
- See that the decorated activity cancels execution
- See that the process now follows the path marked by the BoundaryEvent
- See that the BoundaryEvent instance has produced only one `onEnter` and one `onExit` token

To check the resumption of ProcessInstances:

**Note:** The changes described in the following steps have to be performed manually in your database.

1.
- Run a ProcessInstance and finish it (A ProcessModel with StartEvent -> ScriptTask -> EndEvent is sufficient for this)
- Change the state of the ProcessInstance to running
- Manually remove the FlowNodeInstance of the EndEvent
- Change the state of the ScriptTask to "running" and remove the "onExit" token
- The ProcessInstance will be resumed until the EndEvent is reached

2.
- Run a ProcessInstance and finish it
- Change the state of the ProcessInstance to running
- Manually remove the FlowNodeInstance of the EndEvent
- Do NOT modify any of the other FlowNodeInstances
- The ProcessInstance will be resumed until the EndEvent is reached

3.
- Run a ProcessInstance and finish it
- Change the state of the ProcessInstance to running
- Do NOT modify any of the FlowNodeInstances
- The ProcessInstance will be regared as orphaned. It will be put into a "finished" state
